plugins/targetlocker/dblocker/dblocker.go

```diff
@@ -0,0 +1,310 @@
+	"github.com/facebookincubator/contest/pkg/target"
+	"github.com/facebookincubator/contest/pkg/types"
+
+	// this blank import registers the mysql driver
+	_ "github.com/go-sql-driver/mysql"
+)
+
+// Name is the plugin name.
+var Name = "DBLocker"
+
+var log = logging.GetLogger("targetlocker/" + strings.ToLower(Name))
+
+// dblock represents parts of lock in the database, basically
+// a row from SELECT target_id, job_ID, expires_at
+type dblock struct {
+	targetId  string
+	jobID     int64
+	expiresAt time.Time
+}
+
+// String pretty-prints dblocks for logging and errors
+func (d dblock) String() string {
+	return fmt.Sprintf("target: %s, job %d, expires %s", d.targetId, d.jobID, d.expiresAt)
+}
+
+// targetIDList is a helper to convert contest targets to
+// a list of primary IDs used in the database
+func targetIDList(targets []*target.Target) []string {
+	res := make([]string, 0, len(targets))
+	for _, target := range targets {
+		res = append(res, target.ID)
+	}
+	return res
+}
+
+// listQueryString is a helper to create a (?, ?, ?) string
+// with as many ? as requested
+func listQueryString(length int) string {
+	switch length {
+	case 0:
+		return "()"
+	case 1:
+		return "(?)"
+	default:
+		return "(" + strings.Repeat("?, ", length-1) + "?)"
+	}
+}
+
+// DBLocker implements a simple target locker based on a relational database.
+// The current implementation only supports MySQL officially.
+// All functions in DBLocker are safe for concurrent by multiple goroutines.
+type DBLocker struct {
+	driverName string
+	db         *sql.DB
+	// lockTimeout set on each initial lock request
+	lockTimeout time.Duration
+	// refreshTimeout is used during refresh
+	refreshTimeout time.Duration
+}
+
+// cleanExpired deletes all expired locks on the given targets
+// (lock owner is ignored)
+func (d *DBLocker) cleanExpired(tx *sql.Tx, targets []string, now time.Time) error {
+	q := "DELETE FROM locks WHERE expires_at < ? AND target_id IN " + listQueryString(len(targets)) + ";"
+	// convert targets to a list of interface{}
+	queryList := make([]interface{}, 0, len(targets)+1)
+	queryList = append(queryList, now)
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+	_, err := tx.Exec(q, queryList...)
+	if err != nil {
+		return fmt.Errorf("unable to clean existing locks: %w", err)
+	}
+	return nil
+}
+
+// queryLocks returns a map of ID -> dblock for a given list of targets
+func (d *DBLocker) queryLocks(tx *sql.Tx, targets []string) (map[string]dblock, error) {
+	q := "SELECT target_id, job_id, expires_at FROM locks WHERE target_id IN " + listQueryString(len(targets))
+	// convert targets to a list of interface{}
+	queryList := make([]interface{}, 0, len(targets))
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+
+	rows, err := tx.Query(q, queryList...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read existing locks: %w", err)
+	}
+	defer rows.Close()
```

Comment:
*cough* (just kidding, the docs say that `Close` is idempotent, and users should check `sql.Rows.Err()` instead, as documented at https://golang.org/pkg/database/sql/#Rows )